java8/src/main/java/com/dslplatform/json/processor/EnumTemplate.java

```diff
@@ -22,53 +25,101 @@ private static boolean isAllSimple(StructInfo si) {
 		return true;
 	}
 
-	static void writeName(Context context, AttributeInfo attr, String readValue) throws IOException {
-		StructInfo target = context.structs.get(attr.typeName);
-		if (isAllSimple(target)) {
-			context.code.append("{ writer.writeByte((byte)'\"'); writer.writeAscii(").append(readValue).append(".name()); writer.writeByte((byte)'\"'); }");
+	static void writeName(final Writer code, final StructInfo target, final String readValue) throws IOException {
+		if (target.enumConstantNameSource != null) {
+			String enumConstantNameType = extractReturnType(target.enumConstantNameSource);
+			if ("int".equals(enumConstantNameType)) {
+				code.append("com.dslplatform.json.NumberConverter.serialize(");
+				code.append(readValue).append(".").append(target.enumConstantNameSource.toString());
+				code.append(", writer);\n");
+			} else {
+				code.append("writer.writeString(").append(readValue).append(".").append(target.enumConstantNameSource.toString()).append(");\n");
+			}
+		} else if (isAllSimple(target)) {
+			code.append("writer.writeByte((byte)'\"'); writer.writeAscii(").append(readValue).append(".name()); writer.writeByte((byte)'\"');\n");
 		} else {
-			context.code.append("writer.writeString(value.name());\n");
+			code.append("writer.writeString(value.name());\n");
 		}
 	}
 
 	void create(final StructInfo si, final String className) throws IOException {
-		Set<Integer> hashCodes = new HashSet<>();
-		for(String c : si.constants) {
-			hashCodes.add(StructInfo.calcHash(c));
-		}
-		boolean allSimple = isAllSimple(si);
 		code.append("\tpublic final static class EnumConverter implements com.dslplatform.json.JsonWriter.WriteObject<");
 		code.append(className);
 		code.append(">, com.dslplatform.json.JsonReader.ReadObject<").append(className).append("> {\n");
+		if (si.enumConstantNameSource != null) {
+			String enumConstantNameType = extractReturnType(si.enumConstantNameSource);
+			if ("int".equals(enumConstantNameType)) {

```

Comment:
There is the OptimizedConverters map in context which knows how to process specific types (so we can support more types than just int).
But I though about putting this in if/else instead of loading it up from a map.
Map will cause boxing for primitives ;(

It's fine as is for a first version, though.

Reply:
Yeah, I know there is boxing/unboxing overhead, this was made to optimize String lookup. I think we can make some measurments, and depends on it for enum with few constatnts use simple linear lookup instead of HashMap. Also we can create our own IntMap class to avoid boxing overhead.